app/models/manageiq/providers/ansible_tower/shared/automation_manager/tower_api.rb

```diff
@@ -76,9 +85,7 @@ def update_in_provider(params)
     with_provider_object do |provider_object|
       provider_object.update_attributes!(params)
     end
-    if respond_to?(:refresh_in_provider)
-      refresh_in_provider
-    end
+    self.class.send(:refresh_in_provider_notify, manager_id, params, self)
```

Comment:
you can do `self.class.send(:refresh_in_provider_notify, manager_id, params, self.provider_object)` and so that the `refresh_in_provider_notify` parm list can be `(manager_id, params, provider_object)` and so no more checking in that method.

Reply:
It’s not that easy, but still you’re right, James. Originally I wanted to avoid passing the Tower object, because I considered it not DRY enough and I wanted to re-use the [ConfigurationScriptSource#refresh_in_provider](https://github.com/Glutexo/manageiq-providers-ansible_tower/blob/notify_on_project_update/app/models/manageiq/providers/ansible_tower/shared/automation_manager/configuration_script_source.rb#L60) instance method. But we still use `with_provider_object` in [TowerApi#update_in_provider](https://github.com/Glutexo/manageiq-providers-ansible_tower/blob/927d85a9ce927e0cfb57e6fe528b405949f28767/app/models/manageiq/providers/ansible_tower/shared/automation_manager/tower_api.rb#L81), so we can call the [class method](https://github.com/Glutexo/manageiq-providers-ansible_tower/blob/notify_on_project_update/app/models/manageiq/providers/ansible_tower/shared/automation_manager/configuration_script_source.rb#L25) directly with both the Tower object and the model id. I am still not fully convinced it is great, but it’s less complex. And object type checks are usually ugly, so good for now.